torch/distributed/_sharded_tensor/api.py

```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
         # Barrier for all RPCs to finish on all ranks.
         rpc.api._all_gather(None)
 
+    def gather(
+        self,
+        dst: int = 0,
+        out: Optional[torch.Tensor] = None,
+    ) -> None:
+        """
+        Creates a full :class:`Tensor` on rank `dst` by gathering all sharded tensors.
+
+        The API needs to be called on all ranks in SPMD fashion. All ranks should have
+        the same `dst`. `out` should be a full size zero tensor on `dst` and
+        None on all other ranks.
+
+        Args:
+            dst(int): The rank where full tensor is constructed
+            out (:class `torch.Tensor` optional): The output full tensor. Needs to be provided
+            ONLY on `dst`
+        """
+        my_rank = dist.get_rank(self._process_group)
+        full_size = self.metadata().size
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
+
+        shard_tensors = self.local_shards()
+        world_size = dist.get_world_size(self._process_group)
+
+        gathered_shards = [None] * world_size
+        # all_gather_object is not efficient. will revise this part once
+        # NCCL support for dist.gather() is ready
+        with torch.cuda.device(curr_device):
+            dist.all_gather_object(
+                obj=shard_tensors,
+                object_list=gathered_shards,
+                group=self._process_group,
+            )
+
+        if my_rank == dst:
+            dims = len(full_size)
+            for shards in gathered_shards:
+                assert shards is not None, f"gathered shards cannot be None on {dst}"
+                for shard in shards:
+                    metadata = shard.metadata
+                    tensor = shard.tensor.to(curr_device)
+
+                    out_narrow_view = out
+                    for dim in range(dims):
+                        out_narrow_view = out_narrow_view.narrow(
+                            dim,
+                            metadata.shard_offsets[dim],
+                            metadata.shard_lengths[dim],
+                        ).contiguous()
+
+                    out_narrow_view.add_(tensor)
```

Comment:
I guess the main reason you are asserting `zero` for `out` is that you are adding the gathered content to `out`, I think ideally we can just `copy_` it.

Reply:
I see. thanks